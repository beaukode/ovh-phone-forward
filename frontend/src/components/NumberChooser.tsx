import React from "react";
import {
  List,
  ListItem,
  ListItemText,
  makeStyles,
  Dialog,
  DialogTitle,
  ListItemAvatar,
  Avatar,
  colors,
  TextField,
  Fab,
  Collapse,
} from "@material-ui/core";
import {
  Phone as NumberIcon,
  Add as AddIcon,
  Check as ConfirmIcon,
} from "@material-ui/icons";

const useStyles = makeStyles({
  avatar: {
    backgroundColor: colors.blue[100],
    color: colors.blue[600],
  },
});

type SelectCallback = (number?: string) => void;

interface Props {
  numbers: string[];
  onSelect: SelectCallback;
}

const NumberChooser: React.FC<Props> = ({ numbers, onSelect }) => {
  const classes = useStyles();
  const [custom, setCustom] = React.useState("");

  if (numbers.length === 0) {
    return null;
  }

  const handleListItemClick = (value: string) => {
    onSelect(value);
  };

  const handleCustomChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    setCustom(event.target.value);
  };

  const handleCustomConfirm = () => {
    onSelect(custom);
  };

  const customItem = (
    <>
      <ListItem>
        <ListItemAvatar>
          <Avatar>
            <AddIcon />
          </Avatar>
        </ListItemAvatar>
        <ListItemText
          primary={
            <TextField
              placeholder="Un autre numÃ©ro"
              value={custom}
              onChange={handleCustomChange}
              fullWidth
            />
          }
        />
      </ListItem>
      <Collapse in={Boolean(custom)} timeout="auto" unmountOnExit>
        <ListItem style={{ justifyContent: "center" }}>
          <Fab
            color="primary"
            size="medium"
            aria-label="Confirmer"
            onClick={handleCustomConfirm}
          >
            <ConfirmIcon />
          </Fab>
        </ListItem>
      </Collapse>
    </>
  );

  return (
    <Dialog
      open={true}
      onClose={() => onSelect(undefined)}
      aria-labelledby="numberchooser-dialog-title"
    >
      <DialogTitle id="numberchooser-dialog-title">Transfer vers</DialogTitle>
      <List>
        {numbers.map(number => (
          <ListItem
            button
            onClick={() => handleListItemClick(number)}
            key={number}
          >
            <ListItemAvatar>
              <Avatar className={classes.avatar}>
                <NumberIcon />
              </Avatar>
            </ListItemAvatar>
            <ListItemText primary={number} />
          </ListItem>
        ))}
        {customItem}
      </List>
    </Dialog>
  );
};

export default NumberChooser;
