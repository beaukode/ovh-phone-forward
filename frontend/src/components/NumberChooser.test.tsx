import React from "react";
import {
  render,
  fireEvent,
} from "@testing-library/react";
import NumberChooser from "./NumberChooser";

const numbers = ["number-1", "number-2"];
const handleSelect = jest.fn();

describe("<NumberChooser />", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  test("Render", async () => {
    const { getByRole } = render(
      <NumberChooser numbers={numbers} onSelect={handleSelect} />
    );
    expect(getByRole("dialog")).toMatchSnapshot();
  });

  test("Render without number don't show dialog", async () => {
    const { queryByRole } = render(
      <NumberChooser numbers={[]} onSelect={handleSelect} />
    );
    expect(queryByRole("dialog")).not.toBeInTheDocument();
  });

  test("Click on item call onSelect with number", async () => {
    const { getByText } = render(
      <NumberChooser numbers={numbers} onSelect={handleSelect} />
    );
    fireEvent.click(getByText("number-1"));
    expect(handleSelect).toBeCalledTimes(1);
    expect(handleSelect).toBeCalledWith("number-1");
  });

  test("Show confirm button when custom number is filled", async () => {
    const { getByPlaceholderText, queryByLabelText } = render(
      <NumberChooser numbers={numbers} onSelect={handleSelect} />
    );

    const input = getByPlaceholderText("Un autre numéro");
    expect(input).toBeInTheDocument();
    expect(queryByLabelText("Confirmer")).not.toBeInTheDocument();

    fireEvent.change(input, { target: { value: "newnumber-3" } });
    expect(queryByLabelText("Confirmer")).toBeInTheDocument();
  });

  test("Click on confirm call onSelect with custom number", async () => {
    const { getByPlaceholderText, getByLabelText } = render(
      <NumberChooser numbers={numbers} onSelect={handleSelect} />
    );

    fireEvent.change(getByPlaceholderText("Un autre numéro"), {
      target: { value: "newnumber-3" },
    });
    fireEvent.click(getByLabelText("Confirmer"));

    expect(handleSelect).toBeCalledTimes(1);
    expect(handleSelect).toBeCalledWith("newnumber-3");
  });
});
