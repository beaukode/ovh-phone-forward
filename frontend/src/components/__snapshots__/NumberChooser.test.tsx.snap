// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NumberChooser /> Render 1`] = `
<div
  aria-labelledby="numberchooser-dialog-title"
  class="MuiPaper-root MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthSm MuiPaper-elevation24 MuiPaper-rounded"
  role="dialog"
>
  <div
    class="MuiDialogTitle-root"
    id="numberchooser-dialog-title"
  >
    <h2
      class="MuiTypography-root MuiTypography-h6"
    >
      Transfer vers
    </h2>
  </div>
  <ul
    class="MuiList-root MuiList-padding"
  >
    <div
      aria-disabled="false"
      class="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-button"
      role="button"
      tabindex="0"
    >
      <div
        class="MuiListItemAvatar-root"
      >
        <div
          class="MuiAvatar-root MuiAvatar-circle makeStyles-avatar-1 MuiAvatar-colorDefault"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M6.62 10.79c1.44 2.83 3.76 5.14 6.59 6.59l2.2-2.2c.27-.27.67-.36 1.02-.24 1.12.37 2.33.57 3.57.57.55 0 1 .45 1 1V20c0 .55-.45 1-1 1-9.39 0-17-7.61-17-17 0-.55.45-1 1-1h3.5c.55 0 1 .45 1 1 0 1.25.2 2.45.57 3.57.11.35.03.74-.25 1.02l-2.2 2.2z"
            />
          </svg>
        </div>
      </div>
      <div
        class="MuiListItemText-root"
      >
        <span
          class="MuiTypography-root MuiListItemText-primary MuiTypography-body1"
        >
          number-1
        </span>
      </div>
      <span
        class="MuiTouchRipple-root"
      />
    </div>
    <div
      aria-disabled="false"
      class="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-button"
      role="button"
      tabindex="0"
    >
      <div
        class="MuiListItemAvatar-root"
      >
        <div
          class="MuiAvatar-root MuiAvatar-circle makeStyles-avatar-1 MuiAvatar-colorDefault"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M6.62 10.79c1.44 2.83 3.76 5.14 6.59 6.59l2.2-2.2c.27-.27.67-.36 1.02-.24 1.12.37 2.33.57 3.57.57.55 0 1 .45 1 1V20c0 .55-.45 1-1 1-9.39 0-17-7.61-17-17 0-.55.45-1 1-1h3.5c.55 0 1 .45 1 1 0 1.25.2 2.45.57 3.57.11.35.03.74-.25 1.02l-2.2 2.2z"
            />
          </svg>
        </div>
      </div>
      <div
        class="MuiListItemText-root"
      >
        <span
          class="MuiTypography-root MuiListItemText-primary MuiTypography-body1"
        >
          number-2
        </span>
      </div>
      <span
        class="MuiTouchRipple-root"
      />
    </div>
    <li
      class="MuiListItem-root MuiListItem-gutters"
    >
      <div
        class="MuiListItemAvatar-root"
      >
        <div
          class="MuiAvatar-root MuiAvatar-circle MuiAvatar-colorDefault"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
            />
          </svg>
        </div>
      </div>
      <div
        class="MuiListItemText-root"
      >
        <span
          class="MuiTypography-root MuiListItemText-primary MuiTypography-body1"
        >
          <div
            class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
          >
            <div
              class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiInput-input"
                placeholder="Un autre numÃ©ro"
                type="text"
                value=""
              />
            </div>
          </div>
        </span>
      </div>
    </li>
  </ul>
</div>
`;
